structs-in-rust/article.md

```diff
@@ -0,0 +1,443 @@
+
+Let's add the manual implementation of the `Display`:
+
+```
+impl std::fmt::Display for Point {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "({}, {})", self.x, self.y)
+    }
+}
+
+let p = Point::new(0, 1);
+
+// format a pair
+assert_eq!(format!("{:?}", p), String::from("Point { x: 0, y: 0 }"));
+assert_eq!(format!("{}", p), String::from("(0, 0)"));
+```
+
+The syntax seems pretty natural. We `impl`ement the trait `std::fmt::Display` `for` the struct called `Point`. In the opened code block we then have to define every function that is declared in the trait. In our case, there are only one — `fmt`.
+
+## Struct methods
+
+Just like in Java or C++, in Rust, you can define **methods** for a particular type. This is done in an `impl` block.
+
+```
+impl Point {
+    // Associated function
+    fn new(x: i32, y: i32) -> Point {
+        Point { x, y }
+    }
+
+    // Immutable method
+    fn has_same_x(&self, other: &Self) -> bool {
+        self.x == other.x
+    }
+
+    // Mutable method
+    fn shift_right(&mut self, dx: i32) {
+        self.x += dx;
+    }
+}
+
+// Mutable binding
+let mut point = Point::new(25, 25);
+assert!(!point.has_same_x(&Point::new(30, 25)));
+
+// Calling a method that changes the object state
+point.shift_right(3);
+assert_eq!(point.x, 28);
+```
+
+There are two types of functions related to a particular type.
+
+- Associated functions. They don't require an instance to be called on. Like static functions in Java.
+- Methods. They are invoked on an instance of a type.
+
+### Self
+
+You may have noticed that methods have `self` as their first argument. It represents the instance of the struct the method is being called on, similar to how it is done in Python. Some syntax sugar is involved here. Let's desugar it in two steps.
+
+|     |                                              |                                             |
+| --- | -------------------------------------------- | ------------------------------------------- |
+| 1   | `fn has_same_x(&self, other: &Self)`         | `fn shift_right(&mut self, dx: i32)`        |
+| 2   | `fn has_same_x(self: &Self, other: &Self)`   | `fn shift_right(self: &mut Self, dx: i32)`  |
+| 3   | `fn has_same_x(self: &Point, other: &Point)` | `fn shift_right(self: &mut Point, dx: i32)` |
+
+
+`Self` is an alias for the type of the impl block. In our case, it is `Point`.
+
+Methods can mutate the struct instance they are associated with, but this requires `&mut self` as the first argument. This way they can be called only via mutable binding.
+
+### Why use methods instead of regular functions?
+
+There are at least two reasons.
+
+1. Nice-looking dot notation.
+2. Code organization. It is a lot more convenient for the user of your library to see all functionality of the struct inside one impl block instead of tens of functions spread all across the module.
+
+## Methods vs traits
```

Comment:
I dunno, is it worth going into this in a beginner-level post?

A beginner is unlikely to create their own traits, so the rule is simple:
* if you are implementing a common functionality for which a trait exists (converting to string, comparison, ...), maybe `impl` that trait
* if you are doing something specific to your application, probably use methods in regular `impl` blocks

Reply:
I think the rule is good and can be added.

The rest of it should probably stay, as much for beginners as for SEO.

Reply:
Added